src/daemon/utils_cache_test.c

```diff
@@ -0,0 +1,128 @@
+/**
+ * collectd - src/daemon/utils_cache_test.c
+ * Copyright (C) 2024       Florian octo Forster
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ **/
+
+#include "collectd.h"
+
+#include "daemon/utils_cache.h"
+#include "testing.h"
+#include "utils/common/common.h"
+
+DEF_TEST(uc_get_rate) {
+  struct {
+    char const *name;
+    value_t first_value;
+    value_t second_value;
+    cdtime_t first_time;
+    cdtime_t second_time;
+    metric_type_t type;
+
+    gauge_t want;
+  } cases[] = {
+      {
+          .name = "gauge",
+          .first_value = (value_t){.gauge = 1.0},
+          .second_value = (value_t){.gauge = 2.0},
+          .first_time = TIME_T_TO_CDTIME_T(100),
+          .second_time = TIME_T_TO_CDTIME_T(110),
+          .type = METRIC_TYPE_GAUGE,
+          .want = 2.0,
```

Comment:
To verify gauge handles decreasing values differently:
```suggestion
          .name = "gauge",
          .first_value = (value_t){.gauge = (double)UINT64_MAX-2.0},
          .second_value = (value_t){.gauge = 1.0},
          .first_time = TIME_T_TO_CDTIME_T(100),
          .second_time = TIME_T_TO_CDTIME_T(110),
          .type = METRIC_TYPE_GAUGE,
          .want = 1.0,
```

Reply:
Good idea, added.